### PR TITLE
Decouple generation of genome indices

### PR DESCRIPTION
**Pull request by jkokosar**
## Checklist

* [ ] Update CHANGELOG.rst for each commit separately:
  * Pay attention to write entries under the "Unreleased" section.
  * Mark all breaking changes as "**BACKWARD INCOMPATIBLE:**" and put them
    before non-breaking changes.
  * If a commit modifies a feature listed under "Unreleased" section,
    it might be sufficient to modify the existing CHANGELOG entry from previous
    commit(s).
* [ ] Bump the process version:
  * **MAJOR version (first number)**: Backward incompatible changes (changes
    that break the api/interface). Examples: renaming the input/output, adding
    mandatory input, removing input/output...
  * **MINOR version (middle number)**: add functionality or changes in a
    backwards-compatible manner. Examples: add output field, add non-mandatory
    input parameter, use a different tool that produces same results...
  * **PATCH version (last number)**: changes/bug fixes that do not affect
    the api/interface. Examples: typo fix, change/add warning messages...
* [ ] All inputs are used in process.
* [ ] All output fields have their ``re-save`` calls.

### Processes ###

* Code that was once accepted should be kept like it is. It should be changed
  only when its functionality has changed. Small parts may be changed
  to fix bugs (by changing code only to make it nicer you might introduce a bug).
* Set the optimal number of cores that the tool can use (check if parallelization
  can be performed).
* Description of tool should be self explanatory with links to tool's homepage
  and publication.
* Write explanatory input labels and add short command line parameter names
  in square brackets (for example: Maximum length [``--maxlength``]).
* Where possible provide default values for the process input fields.
  If they are provided, then ``required: false`` is not needed.
* Add descriptive error and warning messages.
* Place a ``re-checkrc`` call after each process step (call of a tool, a
  Bash command, a Python script or an R script).
* Add ``re-progress`` calls where appropriate.

### Commit and PR messages ###

* Each commit should be minimal (i.e. 1 change) and self-contained (including
  tests).
* Changes to Docker images need to be accepted in a separate pull request.
  Before the new images can be used for automated testing of your PR, they
  must be manually tagged and pushed to Docker Hub.
* Mark incomplete PRs with the [WIP] (Work In Progress) tag.
* Do not paste links to private repositories from the public one.

### Tests ###

* Add meaningful names of test files with spaces:
  * Genomes: ``genome species.fa.gz``
  * Annotations: ``annotation species.gtf.gz``
  * Adapters: ``adapters-name/type-source``.
* Output and specific input files should include the name of the tested tool.
* Keep test files' size small. If there is no other option, add the large
  test file to the ``files/large`` directory so it will be tracked with Git LFS.
* Pay attention to edge cases: what will happen if any data is
  missing, if length of some list is 0, if string is given where one expects
  integer, if file is empty, ...
* When testing workflows, try to verify workflow success by checking one
  of the last steps. Pick the one that is most likely to fail. Also test
  that all steps have 'OK' status.

### CODE REVIEW

File: resolwe_bio/tests/processes/test_junctions.py [hunk 1]
```diff
@@ -76,28 +76,28 @@ def test_no_sj_inputs(self):
...
-            inputs = {
+            star_no_junctions = self.run_process('alignment-star', {
                 'genome': star_index_no_junctions.id,
                 'reads': reads_no_junctions.id,
-            }
-            star_no_junctions = self.run_process('alignment-star', inputs)
+            })
 
-            inputs = {
+            empty_bam_upload = self.run_process('upload-bam', {
```
Comment (JureZmrzlikar) [inline resolwe_bio/tests/processes/test_junctions.py]: It would be best to put such formatting in a separate commit, but IMO your time is better spent on other stuff than making this "right". Just for future reference. :-)

File: resolwe_bio/processes/WGBS/walt_index.py [hunk 2]
```diff
@@ -0,0 +1,73 @@
...
+    class Input:
+        """Input fields for WaltIndex."""
+
+        ref_seq = DataField('seq:nucleotide', label='Reference sequence (nucleotide FASTA)')
+
+    class Output:
+        """Output fields to process WaltIndex."""
+
+        index = DirField(label='WALT index')
+        fastagz = FileField(label='FASTA file (compressed)')
```
Comment (JureZmrzlikar) [inline resolwe_bio/processes/WGBS/walt_index.py]: Wy do we need fasta and fastagz outputs here? Isn't is ufficient to have them in the FASTA Data object? This duplicates the files and also disk usage...

Comment (jkokosar) [reply resolwe_bio/processes/WGBS/walt_index.py]: I agree that this is not efficient in terms of storage. However:
- none of the currently supported aligners need .fasta for the alignment job. However, this might not always be the case. It's easier / less costly to have the input fasta next to the index now, than to handle this in some future migration.
- A typical sample will now have an index object on the input. Consider a scenario, where you want to see the resulting .bam in a genome browser. For this you also need the reference sequence. If the .fasta is not attached to the index object, one would need to fetch a reference sequence through additional parent/child relations.
- Existing workflows often take a current ``data:genome:fasta`` on the input. Often, both the index and the fasta are used in (multiple) workflow steps. By keeping the .fasta next to the index, we can retain only one "genome" input in most of the workflows.
- The number of actual index objects is relatively small - the storage problem is not that great in practice.

File: resolwe_bio/processes/import_data/nucleotide_seq.py [hunk 3]
```diff
@@ -0,0 +1,133 @@
...
+    persistence = Persistence.RAW
+    requirements = {
+        'expression-engine': 'jinja',
+        'executor': {
+            'docker': {
+                'image': 'resolwebio/rnaseq:4.9.0'
+            }
+        },
+        'resources': {
+            'cores': 2,
```
[the remark this answers is not in the document]
Comment (jkokosar) [reply resolwe_bio/processes/import_data/nucleotide_seq.py]: I have retained what was in the now retired .yml process. Not sure about the exact reason. We do have a pigz step in the process, and at least that step is significantly faster if multiple cores are available.

**Comment by jkokosar** [comment]
> Have you considered to use more than one core for index building in the processes that currently juts use one core?

Not all of the index building processes benefit from multiple threads. I have retained >1 core only where actually needed.

File: resolwe_bio/processes/WGBS/walt_index.py (hunk 2 above)
Comment (JureZmrzlikar) [reply resolwe_bio/processes/WGBS/walt_index.py]: OK, thanks for the explanation! Agree with you on all points, this makes sense.

File: docs/CHANGELOG.rst [hunk 4]
```diff
@@ -9,10 +9,22 @@ This project adheres to `Semantic Versioning <http://semver.org/>`_.
...
+Added
+-----
+- Add processes: ``bowtie-index``, ``bowtie2-index``, ``bwa-index``,
+  ``hisat2-index``, ``subread-index`` and ``walt-index``.
+
 Changed
 -------
+- **BACKWARD INCOMPATIBLE:** Refactor ``upload-fasta-nucl`` process:
+  ``species`` and ``build`` input information on FASTA file upload  is
+  now mandatory, while ``source`` input is removed.
```
Comment (romunov) [inline docs/CHANGELOG.rst]: is -> has been
